### DISCUSSION TITLE
OO: Dúvida sobre assinatura de métodos e mapeamento e ordem de parâmetros

### DISCUSSION THREAD

**Discussion by oimarcella** [balta-io/forum#550] [Tira Dúvidas] [answered]
Curso: Acesso à dados com .NET, C# ...
Aula: Imersão / Select IN

-- Explicando o contexto:
Minha dúvida surgiu a partir dessa aula onde ele usa o select in numa query e ao invés de colocar um id fixo, ele usa os parâmetros do SQL logo após a o comando sql. 
![image](https://github.com/balta-io/forum/assets/43733159/682434df-fb60-4d97-9113-5d05ab66020f)

Em outras aulas do curso ele usa a função Query da seguinte maneira:
`connection.Query<CarreerItem, Course, CarreerItem>(commandSql, ()=>{//uma funcao}, splotOn:"Id")`;

Nessa aula ele fez:
`connection.Query(commandSql, new {paramentros aqui})`;

-- Finalmente:

Dúvida: **Como eu sei em que ordem passarei os parâmetros da função? Porque as vezes o 2 parâmetro é uma função, e em outras horas ele é um objeto. (Usando como exemplo essa função Query)**

Em funções mais simples criadas por nós mesmos quando eu crio `minhaFuncao(string nome, int idade)` eu não posso passar parâmetros de outro tipo que não seja int, no lugar onde seria para receber idade.

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @mlasa, como vai?

**_Em outras aulas do curso ele usa a função Query da seguinte maneira:
connection.Query<CarreerItem, Course, CarreerItem>(commandSql, ()=>{//uma funcao}, splotOn:"Id");_**
Sempre precisamos passar um valor único como parâmetro... no caso da função, se ela retorna um valor único (inteiro, string, etc) ela serve como parâmetro também.

**_Nessa aula ele fez:
connection.Query(commandSql, new {paramentros aqui});_**
Neste modelo já usamos variáveis como parâmetros ou valores fixos, o que também é possível.

 **_Como eu sei em que ordem passarei os parâmetros da função? Porque as vezes o 2 parâmetro é uma função, e em outras horas ele é um objeto. (Usando como exemplo essa função Query)_**
Os parâmetros no dapper sempre serão um objeto! Por exemplo:
```csharp
new { Id = 1 }
```

Se precisar passar mais parâmetros:
```csharp
new { Id = 1, Nome = "andré" }
```

E qual ordem deles?
Não tem! Para cada parâmetro no seu SQL (Exemplo: `SELECT * FROM [Tabela] WHERE Id=@id` -- `@id` é um parâmetro) ele vai buscar uma propriedade do objeto com mesmo nome.

Ficou mais claro?

**Reply by oimarcella** [reply]
Huuum siiim, fez até um clique na minha mente!

Mas tenho outra dúvida, se os parâmetros para o Dapper são sempre objetos, oque seria o splitOn nesse trecho?
`connection.Query<CarreerItem, Course, CarreerItem>(commandSql, ()=>{//uma funcao}, splitOn:"Id");`
Ele é como se fosse só uma das propriedades de um objeto, não um objeto inteiro.

**Reply by henriqueweege** [reply]
Oi, @mlasa !
Acho que essa consigo te responder!
Os dados que vem do banco não vem separados por tabela/entidade, mas sim todos juntos, na ordem estabelecida na query. Por conta disso, precisamos dizer ao Dapper onde termina um objeto e onde começa outro. Essa é a função do "splitOn".

**Reply by oimarcella** [reply]
Olá @henriqueweege eu quis dizer algo diferente com a pergunta.

Quando passo SplitOn: "Id", não estou passando um objeto, certo?
Qual seria esse tipo de parâmetro?

Porque se fosse objeto seria

`new{
   SplitOn: "Id"
}`

**Reply by henriqueweege** [reply]
Então, se estiver usando VS Studio, basta deixar o cursor em cima do nome do método e apertar a tecla F12 para ir para a definição do mesmo. 
Olhando os parâmetros que esse método recebe, tem vários que são opcionais, inclusive o "splitOn". O tipo dele é uma string mesmo, mas é necessário explicitar qual é o parâmetro sendo passado por terem vários opcionais.

![image](https://github.com/balta-io/forum/assets/96916801/2338ce43-06d1-4f31-ba83-106d80353465)


Vou deixar um exemplo aqui de um método que criei que tem essa mesma proposta de ter vários parâmetros opcionais.

![image](https://github.com/balta-io/forum/assets/96916801/ef076669-7b78-46e4-9846-8d6c2d3a8e3a)

Se você implementar esse exemplo e rodar ele, vai perceber que na primeira chamada ele cai no segundo if, e na segunda chamada ele cai no primeiro if, por conta da validação se o argumento é null ou vazio.

Espero que ajude!

Valeeu!
